### ISSUE TITLE
FeatureLinkerUnabeled[QT] doesn't work in TOPPAS

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#2194]
Problem originally reported by Mike Rightmire on the mailing list.

FeatureLinkerUnlabeled and FeatureLinkerUnlabeledQT need a list of multiple files as input. In TOPPAS they must be preceded by a Collector node for this reason (unless the previous node already produces a file list as output, e.g. MapAligner[...]). However, doing this currently produces the following error:
> ERROR: FeatureLinkerUnlabeledQT failed!
> Did not find upstream nodes with un-recycled names. Something is fishy!

This can be reproduced using a minimal workflow (any test files should do, e.g. from the TOPP tool tests): Input files -> Collect -> FeatureLinkerUnlabeled[QT] -> Output files

I've observed this bug on both Linux (Ubuntu 12.04/14.04) and Mac OS X (10.11.6) with a recent "develop" build.

Mike reported that it worked for him on OS X with the OpenMS 2.0.1 release version.

**Comment by PayamEmami** [comment]
Just tried it with the OpenMS 2.1.0 ([nightly binaries](http://ftp.mi.fu-berlin.de/pub/OpenMS/nightly_binaries/)) on Windows and it worked without any errors.

**Comment by BioComSoftware** [comment]
<?xml version="1.0" encoding="ISO-8859-1"?>
<PARAMETERS version="1.6.2" xsi:noNamespaceSchemaLocation="http://open-ms.sourceforge.net/schemas/Param_1_6_2.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <NODE name="info" description="">
    <ITEM name="version" value="2.1.0" type="string" description="" required="false" advanced="false" />
    <ITEM name="num_vertices" value="4" type="int" description="" required="false" advanced="false" />
    <ITEM name="num_edges" value="3" type="int" description="" required="false" advanced="false" />
    <ITEM name="description" value="&lt;![CDATA[]]&gt;" type="string" description="" required="false" advanced="false" />
  </NODE>
  <NODE name="vertices" description="">
    <NODE name="2" description="">
      <ITEM name="recycle_output" value="false" type="string" description="" required="false" advanced="false" />
      <ITEM name="toppas_type" value="tool" type="string" description="" required="false" advanced="false" />
      <ITEM name="tool_name" value="FeatureLinkerUnlabeledQT" type="string" description="" required="false" advanced="false" />
      <ITEM name="tool_type" value="" type="string" description="" required="false" advanced="false" />
      <ITEM name="x_pos" value="40" type="double" description="" required="false" advanced="false" />
      <ITEM name="y_pos" value="20" type="double" description="" required="false" advanced="false" />
      <NODE name="parameters" description="">
        <ITEMLIST name="in" type="input-file" description="input files separated by blanks" required="true" advanced="false" supported_formats="*.featureXML,*.consensusXML">
        </ITEMLIST>
        <ITEM name="out" value="" type="output-file" description="Output file" required="true" advanced="false" supported_formats="*.consensusXML" />
        <ITEM name="keep_subelements" value="false" type="string" description="For consensusXML input only: If set, the sub-features of the inputs are transferred to the output." required="false" advanced="false" restrictions="true,false" />
        <ITEM name="log" value="" type="string" description="Name of log file (created only when specified)" required="false" advanced="true" />
        <ITEM name="debug" value="0" type="int" description="Sets the debug level" required="false" advanced="true" />
        <ITEM name="threads" value="1" type="int" description="Sets the number of threads allowed to be used by the TOPP tool" required="false" advanced="false" />
        <ITEM name="no_progress" value="true" type="string" description="Disables progress logging to command line" required="false" advanced="false" restrictions="true,false" />
        <ITEM name="force" value="false" type="string" description="Overwrite tool specific checks." required="false" advanced="true" restrictions="true,false" />
        <ITEM name="test" value="false" type="string" description="Enables the test mode (needed for internal use only)" required="false" advanced="true" restrictions="true,false" />
        <NODE name="algorithm" description="Algorithm parameters section">
          <ITEM name="use_identifications" value="false" type="string" description="Never link features that are annotated with different peptides (only the best hit per peptide identification is taken into account)." required="false" advanced="false" restrictions="true,false" />
          <ITEM name="nr_partitions" value="1" type="int" description="How many partitions in m/z space should be used for the algorithm (more partitions means faster runtime and more memory efficient execution )" required="false" advanced="false" restrictions="1:" />
          <ITEM name="ignore_charge" value="false" type="string" description="false [default]: pairing requires equal charge state (or at least one unknown charge &apos;0&apos;); true: Pairing irrespective of charge state" required="false" advanced="false" restrictions="true,false" />
          <NODE name="distance_RT" description="Distance component based on RT differences">
            <ITEM name="max_difference" value="100" type="double" description="Never pair features with a larger RT distance (in seconds)." required="false" advanced="false" restrictions="0:" />
            <ITEM name="exponent" value="1" type="double" description="Normalized RT differences ([0-1], relative to &apos;max_difference&apos;) are raised to this power (using 1 or 2 will be fast, everything else is REALLY slow)" required="false" advanced="true" restrictions="0:" />
            <ITEM name="weight" value="1" type="double" description="Final RT distances are weighted by this factor" required="false" advanced="true" restrictions="0:" />
          </NODE>
          <NODE name="distance_MZ" description="Distance component based on m/z differences">
            <ITEM name="max_difference" value="0.3" type="double" description="Never pair features with larger m/z distance (unit defined by &apos;unit&apos;)" required="false" advanced="false" restrictions="0:" />
            <ITEM name="unit" value="Da" type="string" description="Unit of the &apos;max_difference&apos; parameter" required="false" advanced="false" restrictions="Da,ppm" />
            <ITEM name="exponent" value="2" type="double" description="Normalized ([0-1], relative to &apos;max_difference&apos;) m/z differences are raised to this power (using 1 or 2 will be fast, everything else is REALLY slow)" required="false" advanced="true" restrictions="0:" />
            <ITEM name="weight" value="1" type="double" description="Final m/z distances are weighted by this factor" required="false" advanced="true" restrictions="0:" />
          </NODE>
          <NODE name="distance_intensity" description="Distance component based on differences in relative intensity (usually relative to highest peak in the whole data set)">
            <ITEM name="exponent" value="1" type="double" description="Differences in relative intensity ([0-1]) are raised to this power (using 1 or 2 will be fast, everything else is REALLY slow)" required="false" advanced="true" restrictions="0:" />
            <ITEM name="weight" value="0" type="double" description="Final intensity distances are weighted by this factor" required="false" advanced="true" restrictions="0:" />
          </NODE>
        </NODE>
      </NODE>
    </NODE>
    <NODE name="1" description="">
      <ITEM name="recycle_output" value="false" type="string" description="" required="false" advanced="false" />
      <ITEM name="toppas_type" value="merger" type="string" description="" required="false" advanced="false" />
      <ITEM name="x_pos" value="29.3333333333333" type="double" description="" required="false" advanced="false" />
      <ITEM name="y_pos" value="-146.666666666667" type="double" description="" required="false" advanced="false" />
      <ITEM name="round_based" value="false" type="string" description="" required="false" advanced="false" />
    </NODE>
    <NODE name="0" description="">
      <ITEM name="recycle_output" value="false" type="string" description="" required="false" advanced="false" />
      <ITEM name="toppas_type" value="input file list" type="string" description="" required="false" advanced="false" />
      <ITEMLIST name="file_names" type="string" description="" required="false" advanced="false">
        <LISTITEM value="../../../../usr/local/OpenMS/doc/code_examples/data/Tutorial_Unlabeled_1.featureXML"/>
        <LISTITEM value="../../../../usr/local/OpenMS/doc/code_examples/data/Tutorial_Unlabeled_2.featureXML"/>
      </ITEMLIST>
      <ITEM name="x_pos" value="20" type="double" description="" required="false" advanced="false" />
      <ITEM name="y_pos" value="-300" type="double" description="" required="false" advanced="false" />
    </NODE>
    <NODE name="3" description="">
      <ITEM name="recycle_output" value="false" type="string" description="" required="false" advanced="false" />
      <ITEM name="toppas_type" value="output file list" type="string" description="" required="false" advanced="false" />
      <ITEM name="x_pos" value="41.3333333333333" type="double" description="" required="false" advanced="false" />
      <ITEM name="y_pos" value="164" type="double" description="" required="false" advanced="false" />
      <ITEM name="output_folder_name" value="" type="string" description="" required="false" advanced="false" />
    </NODE>
  </NODE>
  <NODE name="edges" description="">
    <NODE name="0" description="">
      <NODE name="source/target" description="">
        <ITEM name="" value="0/1" type="string" description="" required="false" advanced="false" />
      </NODE>
      <NODE name="source_out_param" description="">
        <ITEM name="" value="__no_name__" type="string" description="" required="false" advanced="false" />
      </NODE>
      <NODE name="target_in_param" description="">
        <ITEM name="" value="__no_name__" type="string" description="" required="false" advanced="false" />
      </NODE>
    </NODE>
    <NODE name="1" description="">
      <NODE name="source/target" description="">
        <ITEM name="" value="1/2" type="string" description="" required="false" advanced="false" />
      </NODE>
      <NODE name="source_out_param" description="">
        <ITEM name="" value="__no_name__" type="string" description="" required="false" advanced="false" />
      </NODE>
      <NODE name="target_in_param" description="">
        <ITEM name="" value="in" type="string" description="" required="false" advanced="false" />
      </NODE>
    </NODE>
    <NODE name="2" description="">
      <NODE name="source/target" description="">
        <ITEM name="" value="2/3" type="string" description="" required="false" advanced="false" />
      </NODE>
      <NODE name="source_out_param" description="">
        <ITEM name="" value="out" type="string" description="" required="false" advanced="false" />
      </NODE>
      <NODE name="target_in_param" description="">
        <ITEM name="" value="__no_name__" type="string" description="" required="false" advanced="false" />
      </NODE>
    </NODE>
  </NODE>
</PARAMETERS>

**Comment by timosachsenberg** [comment]
I think this was fixed in https://github.com/OpenMS/OpenMS/issues/2166

**Comment by BioComSoftware** [comment]
@timosachsenberg 

Hi Tim, 

I'm not sure its the same issue. That ticket was closed 17 days ago. But we're using a new install (installed just today).

**Comment by hendrikweisser** [comment]
#2180 was only merged into "master". @BioComSoftware, you are using the "develop" branch, correct?

**Comment by BioComSoftware** [comment]
Sorry I don't have an answer. I'll have to check with the admin who did the install.

**Comment by timosachsenberg** [comment]
Hi,
look above - it says: Git branch : develop
this develop branch doesn't yet contain the fix.
Only after: https://github.com/OpenMS/OpenMS/pull/2197 is merged (hopefully soon)
Best
